pkg/sql/sqlbase/testutils.go

```diff
@@ -1473,6 +1552,103 @@ func randIndexTableDefFromCols(
 	return tree.IndexTableDef{Columns: indexElemList}
 }
 
+// randPartialIndexPredicateFromCols creates a partial index expression with
+// a random subset of the given columns.
+func randPartialIndexPredicateFromCols(
+	rng *rand.Rand, columnTableDefs []*tree.ColumnTableDef, tableName *tree.TableName,
+) tree.Expr {
+	// Shuffle the columns.
+	cpy := make([]*tree.ColumnTableDef, len(columnTableDefs))
+	copy(cpy, columnTableDefs)
+	rng.Shuffle(len(cpy), func(i, j int) { cpy[i], cpy[j] = cpy[j], cpy[i] })
+
+	// Select a random number of columns (at least 1). Loop through the columns
+	// to find columns with types that are currently supported for generating
+	// partial index expressions.
+	nCols := rng.Intn(len(cpy)) + 1
+	cols := make([]*tree.ColumnTableDef, 0, nCols)
+	for _, col := range cpy {
+		if isAllowedPartialIndexColType(col) {
+			cols = append(cols, col)
+		}
+		if len(cols) == nCols {
+			break
+		}
+	}
+
+	// Build a boolean expression tree with containing a reference to each
+	// column.
+	var e tree.Expr
+	for _, columnTableDef := range cols {
+		expr := randBoolColumnExpr(rng, columnTableDef, tableName)
+		// If an expression has already been built, combine the previous and
+		// current expression with an AndExpr or OrExpr.
+		if e != nil {
+			expr = randAndOrExpr(rng, e, expr)
+		}
+		e = expr
+	}
+	return e
+}
+
+// isAllowedPartialIndexColType returns true if the column type is supported and
+// the column can be included in generating random partial index predicate
+// expressions.
+func isAllowedPartialIndexColType(columnTableDef *tree.ColumnTableDef) bool {
+	fam := columnTableDef.Type.(*types.T).Family()
```

Comment:
can this be a type switch instead? also where's the canonical source for this information? can you add a comment for that so it's clearer how to keep this list in sync?

Reply:
I just did a quick scan through the list of "interesting datums" for types that were valid in a comparison operator. It's not exhaustive at this point. I've added a comment to make clear what is supported currently.

A type switch would require listing all types rather than just families. If we were to add a new Int type, we'd have to make a change here to support creating partial index predicates with the new type. With families we don't have to worry about that.

I could do a switch on the family, if that's preferred:

```go
switch columnTableDev.Type(*types.T).Family() {
case types.BoolFamily:
    return true
case types.IntFamily:
    return true
// ...
}

Reply:
```
case types.BoolFamily,
  types.IntFamily,
  ...:
  return true
``` 
is what I had in mind

Reply:
I misunderstood what was going on here about why this family check is even needed.